### PR TITLE
Sigma clip aperture mask

### PR DESCRIPTION
**Pull request by abostroem**
This PR adds a functionality to return a list of tuples corresponding to objects in an aperture object. Each tuple contains 2, 2D arrays: a cutout of the image around a given aperture and a fractional aperture mask.

### CODE REVIEW

**Comment by larrybradley** [comment]
@abostroem Thanks again for working on this during the SciPy sprints!  I'll review this soon.

**Comment by larrybradley** [comment]
@bsipocz, @cdeil, @astrofrog:  Can you please review this as well?  This PR is intended to address #8, #161, and #259.

**Comment by bsipocz** [comment]
@abostroem - Thanks for working on this. I'll try this out in the weekend.

**Comment by abostroem** [comment]
@bsipocz Awesome. Let me a few notes:

* Currently this only support circular apertures but I think it would be pretty easy to add the other ones (or say a sky annulus). 
* For testing I compared the flux from the current aperture photometry to the summed flux from the output of the new function. I was only able to match up to 0.01. We called that good enough, but I haven't hunted down what is different as this code largely draws on the same functions that the aperture photometry uses.

**Comment by bsipocz** [comment]
@abostroem - First of all, thank you for working on this, and I'm really sorry that it took me so long to do the review. As the first comment I wonder whether you have some more time to work on this? Like adding some paragraph to the docs describing it? I'll have some inline comments, too. Some of the are very nitpicky, so if you're short on time we can merge this as I'll fix those later.

File: photutils/mask_data.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    positions = circular_aperture.positions
+    radius = circular_aperture.r
+    extents = np.zeros((len(positions), 4), dtype=int)
+
+    extents[:, 0] = positions[:, 0] - radius + 0.5
+    extents[:, 1] = positions[:, 0] + radius + 1.5
+    extents[:, 2] = positions[:, 1] - radius + 0.5
+    extents[:, 3] = positions[:, 1] + radius + 1.5
+
+    from .aperture_funcs import get_phot_extents
```
Comment (bsipocz) [inline photutils/mask_data.py]: Move the import to the top of the method.  I would even say that you can put in on the top of the file.

File: photutils/mask_data.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import warnings
+import astropy.units as u
+from astropy.utils.exceptions import AstropyUserWarning
+
+
+__all__ = ['get_cutouts']
+
+
+
+def get_cutouts(data, circular_aperture, use_exact = 0, subpixels = 5):
```
Comment (bsipocz) [inline photutils/mask_data.py]: nitpicking: don't put spaces around "="  for kwargs.

File: photutils/mask_data.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    '''
+    Create a square cutout of each aperture and an associated mask which denotes the
+    aperture with 1 for inside, 0 for outside, and a fraction for the edges.
+
+    Parameters:
+    -----------
+    data : Image array
```
Comment (bsipocz) [inline photutils/mask_data.py]: this can be reformatted with adding extra lines between the parameters

File: photutils/mask_data.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    circular_aperture: photutils aperture object
+    use_exact : (default = 0) refers to the keyword in .geometry.circular_overlap_grid.
+            If use_exact = 1, a hard boundary containing only the pixels fully contained
+            within the aperture. if use_exact = 0, calculates the fractional pixel
+            contained within the pixel
+    subpixels : (default = 5) if use_exact = 0 this is used to determine the fraction of
```
Comment (bsipocz) [inline photutils/mask_data.py]: same here, brake the line

File: photutils/mask_data.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            the pixel contained within the aperture.
+
+    Returns:
+    -----------
+    img_mask_list : a list with one tuple for each object position in the circular_aperture
```
Comment (bsipocz) [inline photutils/mask_data.py]: add a line brake

File: photutils/mask_data.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+def get_cutouts(data, circular_aperture, use_exact = 0, subpixels = 5):
+    '''
+    Create a square cutout of each aperture and an associated mask which denotes the
+    aperture with 1 for inside, 0 for outside, and a fraction for the edges.
+
+    Parameters:
+    -----------
+    data : Image array
+    circular_aperture: photutils aperture object
+    use_exact : (default = 0) refers to the keyword in .geometry.circular_overlap_grid.
```
Comment (bsipocz) [inline photutils/mask_data.py]: add a line brake here, so the first one only contains the name of the parameter and the type

**Comment by bsipocz** [comment]
@larrybradley - It seems that there is quite a lot code repetition if we have this feature in a separate file. If there any reason that it cannot live in ``aperture_func`` and have it activated with a new kwarg in ``do_circular_photometry``, etc?

File: photutils/mask_data.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        fractional_overlap_mask = get_fractional_overlap_mask(absolute_extent[indiv_obj],
+                                                                centered_extent[indiv_obj],
+                                                                ~no_overlap[indiv_obj],
+                                                                radius, use_exact,
+                                                                subpixels)
+        img_stamp = cutout_image(data, absolute_extent[indiv_obj], ~no_overlap[indiv_obj])
+        img_mask_list.append((img_stamp, fractional_overlap_mask))
+    return img_mask_list
+
+def get_fractional_overlap_mask(indiv_absolute_extent, indiv_centered_extent, overlap,
```
Comment (bsipocz) [inline photutils/mask_data.py]: I think you can merge this function into ``get_cutouts`` as both are very much for circular case; and support for new aperture classes could be done in new functions to keep the structure clear

**Comment by abostroem** [comment]
@bsipocz Thanks for the comments. I'm currently studying for my Quals (two weeks away). I'll take a look at your comments at the end of the month

**Comment by larrybradley** [comment]
@bsipocz I agree that this code can live in `aperture_func`.  However, it's a separate feature from the actual photometry, so it should probably remain a separate function.

@abostroem Will you have a chance to address @bsipocz's comments soon?  If not, we can merge this PR now and then fix the small issues later.

**Comment by abostroem** [comment]
@larrybradley @bsipocz I'll see what I can get done tonight - the comments looked mostly cosmetic so I should be able to get to everything.

File: photutils/mask_data.py (hunk 1 above)
Comment (abostroem) [inline photutils/mask_data.py]: Done

File: photutils/mask_data.py (hunk 2 above)
Comment (abostroem) [inline photutils/mask_data.py]: Done

File: photutils/mask_data.py (hunk 3 above)
Comment (abostroem) [inline photutils/mask_data.py]: I think I fixed this, can you double check that I made the change you intended?

File: photutils/mask_data.py (hunk 4 above)
Comment (abostroem) [inline photutils/mask_data.py]: I'm not sure I understand, can you provide more details?

File: photutils/mask_data.py (hunk 6 above)
Comment (abostroem) [inline photutils/mask_data.py]: Again, let me know if I did this right

File: photutils/mask_data.py (hunk 7 above)
Comment (abostroem) [inline photutils/mask_data.py]: @bsipocz can you please provide more details on what you envision here? I'm confused about what you want to put into functions and if anything needs to be changed to create a new aperture class.

@larrybradley I can't tell if you think it should stay as is or change?

**Comment by abostroem** [comment]
@larrybradley @bsipocz Thank you for the feedback. I managed to fix most of the formatting issues. I have a few questions about the docstring formatting and the last comment about more substantial structural changes.

File: photutils/mask_data.py (hunk 7 above)
Comment (larrybradley) [inline photutils/mask_data.py]: I think what @bsipocz is suggesting is to combine the `get_fractional_overlap_mask` and `get_cutouts` into a single function.  They are both specific to a `CircularAperture`.

**Comment by larrybradley** [comment]
Thanks again, @abostroem, and good to see you at the AAS meeting!  I'm going to merge this and then fix up a few things.  As I mentioned, I'll likely generalize this for other apertures at some point.

**Comment by mwcraig** [comment]
:tada: This makes me very happy!

**Comment by bsipocz** [comment]
@larrybradley - Not sure about your timescale on this, but I may be able to find time this week to do the fixes. Should have done a while ago...

**Comment by larrybradley** [comment]
@bsipocz Thanks, but I'm actually working on the fixes right now.

**Comment by bsipocz** [comment]
:+1:

**Pull request merged**